### ISSUE TITLE
Demo

### ISSUE THREAD

**Issue by swetha2410** [Axe--/Visual-Question-Answering#4]
Could you give me an idea of how to run the demo of the hierarchical trained model? I am not sure of how to run a trained model.
Thanks in advance

**Comment by Axe--** [comment]
Hey,

I will push the code for inference (demo) in the next of couple of days since I don't have a trained model at the moment.

But here are the steps (& psuedocode) that might be helpful for now:

Before execution, change the args to `--mode test` & `--model_ckpt  your_model_file.ckpt` 

In main.py try adding this near line 285.

```python
    elif args.mode == 'test':
        # Use the same word-index dicts as that obtained for the training set
        test_dataset = VQADataset(args.val_file, args.val_img, word2idx, label2idx, max_seq_length,
                                  transform=Compose([Resize(image_size), ToTensor(), Normalize((0.485, 0.456, 0.406),
                                                                                               (0.229, 0.224, 0.225))]))

        test_loader = torch.utils.data.DataLoader(test_dataset, batch_size, shuffle=True,
                                                  drop_last=True, num_workers=args.num_workers)

        # Num of classes = K + 1 (for UNKNOWN)
        num_classes = args.num_cls + 1

        # Setup model params
        question_encoder_params = model_config['question_params']
        image_encoder_params = model_config['image_params']

        # Define model & load to device
        VQANet = model_config['model']

        model = VQANet(question_encoder_params, image_encoder_params, K=num_classes)
        model.to(device)

        # Load weights from checkpoint file
        log_dir = os.path.join(args.expt_dir, args.expt_name, args.run_name)
        model_ckpt_path = os.path.join(log_dir, args.model_ckpt)

        checkpoint = torch.load(model_ckpt_path, map_location=device)
        model.load_state_dict(checkpoint)

        print('Model successfully loaded from {}'.format(model_ckpt_path))

        test_metrics = compute_validation_metrics(model, test_loader, device, test_dataset.__len__())

        # raise NotImplementedError('TODO: test mode')

```


And if you intend to also compute the predicted labels,
Just modify the `compute_validation_metrics()` as follows:

```python
labels_pred = []      # <----
for i, batch in enumerate(dataloader):
    ...
    label_predicted = torch.argmax(label_logits, dim=1)
    correct = (label == label_predicted)
    num_correct += correct.sum().item()
    labels_pred += label_predicted.detach().cpu().tolist()     # <----
....
metrics['pred'] = labels_pred              # <----
return metrics
```
And then use `idx2word` dict to map the predicted label indices to answers.

**Comment by swetha2410** [comment]
Thanks! 
I just had one more question.
Why does the validation accuracy of the model reduce significantly each time I reload a checkpoint to resume training of the model??
I'm currently trying to train the model in kaggle and can only train for one epoch during one session. But each time I try to resume using the last model (filename.pth) as the checkpoint, the validation accuracy decreases and now during the third epoch, it has come down from 28 % to 2%.

**Comment by Axe--** [comment]
I see, thanks for raising that issue.  
(I had never resumed from checkpoint for this repo)

Sure let me debug that error, and push in a day or two.

Also I will probably modify the checkpoint `state_dict` to store the best validation accuracy:

```python3
# state_dict = model.state_dict()   # currently 
state_dict = {'model_state_dict': model.state_dict(), 
               'optimizer_state_dict': optimizer.state_dict(),
               'curr_step': curr_step, 'loss': loss,
               'epoch': epoch, 'label_set': labels, 
               'val_accuracy': val_acc}
```
So you might consider retraining with the updated code.  :-)
